### DISCUSSION TITLE
The correct way to create users from a JSVM custom route

### DISCUSSION THREAD

**Discussion by charbs-io** [pocketbase/pocketbase#3145] [Q&A] [answered]
I'm implementing a user invite flow.

Fairly simple workflow containing custom js routes. (something like invite user, user receives email, clicks link, token is verified, user sets a password, then invokes a custom endpoint called `/accept`).

In this custom endpoint, i need to delete the old `pendingUser` record, save some user settings and most importantly, create a new user record.

I figured it's a simple operation...

```
const userCollection = $app.dao().findCollectionByNameOrId("users");
const newUserRecord = new Record(userCollection, {
  // email, password, etc 
});
$app.dao().saveRecord(newUserRecord);
```

However, after doing this once, i started getting an error: **GoError: constraint failed: UNIQUE constraint failed: users.tokenKey (2067)**

Looking further into it, I found that if i create users this way, `users.tokenKey` and `users.passwordHash` are simply empty.

I've looked at the JSVM reference and I can't find/see a special method to do this.

**Comment by ganigeorgiev** [comment, marked as answer]
I'll consider documenting it better as this was requested previously too (see https://github.com/pocketbase/pocketbase/issues/3083#issuecomment-1668103101), but user records (and admins) need to use the `setPassword()` helper when setting passwords.

As explained in the above link, the record model doesn't have a `password` field but `passwordHash` and `tokenKey`. The `password` and `passwordConfirm` fields are available only in the `RecordUpsertForm` when loading external data (eg. from a request or any other user input).

If you want to create a user and directly persist it with the Dao, you need to call `setPassword()`, aka.:

```js
const userCollection = $app.dao().findCollectionByNameOrId("users");

const newUserRecord = new Record(userCollection);

newUserRecord.set("email", "test@example.com")

newUserRecord.setPassword("1234567890") // this will take care for generating the bcrypt hash and a random unique tokenKey

$app.dao().saveRecord(newUserRecord);
```

**Reply by charbs-io** [reply]
Thanks a lot @ganigeorgiev .

May i make a suggestion? 

`newUserRecord.setUsername()` should accept an empty paramter - in that case, PocketBase should internally go about creating one using the same convention used under the hood. 

I did 

```
const randomUsername = "users" + Math.floor(Math.random() * 100000);
newUserRecord.setUsername(
    $app.dao().suggestUniqueAuthRecordUsername("users", randomUsername)
);
```

to try and make it conform to a similar standard as the one currently used.
